hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/joiner/DefaultSpanJoinerBuilder.java

```diff
@@ -70,61 +74,122 @@ public Single<SpanJoiner> build(
-          .map(spanIdToSpanMap -> buildSourceToSpanMap(sourceToSpanIdMap, spanIdToSpanMap));
+    @Override
+    public <T> Single<ListMultimap<T, Span>> joinSpans(
+        Collection<T> joinSources, MultipleSpanIdGetter<T> multipleSpanIdGetter) {
+      return this.buildSourceToIdsMap(joinSources, multipleSpanIdGetter)
+          .flatMap(
+              sourceToSpanIdsMap ->
+                  this.buildSpanRequest(sourceToSpanIdsMap, SPANS_KEY)
+                      .flatMap(spanDao::getSpans)
+                      .map(this::buildSpanIdToSpanMap)
+                      .map(
+                          spanIdToSpanMap ->
+                              this.buildSourceToSpanListMultiMap(
+                                  sourceToSpanIdsMap, spanIdToSpanMap)));
     }
 
-    private <T> Map<T, Span> buildSourceToSpanMap(
-        Map<T, String> sourceToSpanIdMap, Map<String, Span> spanIdToSpanMap) {
-      return sourceToSpanIdMap.entrySet().stream()
-          .filter(entry -> spanIdToSpanMap.containsKey(entry.getValue()))
-          .collect(
-              Collectors.toUnmodifiableMap(
-                  Entry::getKey, entry -> spanIdToSpanMap.get(entry.getValue())));
+    private <T> Map<T, Span> reduceMap(Map<T, List<Span>> multiMap) {
+      return multiMap.entrySet().stream()
+          .filter(entry -> !entry.getValue().isEmpty())
+          .collect(Collectors.toUnmodifiableMap(Entry::getKey, entry -> entry.getValue().get(0)));
+    }
+
+    private <T> Single<Map<T, List<String>>> buildSourceToIdMap(
+        Collection<T> joinSources, SpanIdGetter<T> spanIdGetter) {
+      return Observable.fromIterable(joinSources)
+          .flatMapSingle(source -> this.maybeBuildMapEntry(source, spanIdGetter))
+          .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
+    }
+
+    private <T> Single<Map<T, List<String>>> buildSourceToIdsMap(
+        Collection<T> joinSources, MultipleSpanIdGetter<T> multipleSpanIdGetter) {
+      return Observable.fromIterable(joinSources)
+          .flatMapSingle(source -> this.maybeBuildMapEntry(source, multipleSpanIdGetter))
+          .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
+    }
+
+    private <T> Single<Entry<T, List<String>>> maybeBuildMapEntry(
+        T source, SpanIdGetter<T> spanIdGetter) {
+      return spanIdGetter.getSpanId(source).map(List::of).map(ids -> Map.entry(source, ids));
+    }
+
+    private <T> Single<Entry<T, List<String>>> maybeBuildMapEntry(
+        T source, MultipleSpanIdGetter<T> multipleSpanIdGetter) {
+      return multipleSpanIdGetter.getSpanIds(source).map(ids -> Map.entry(source, ids));
+    }
+
+    private <T> ListMultimap<T, Span> buildSourceToSpanListMultiMap(
+        Map<T, List<String>> sourceToSpanIdsMap, Map<String, Span> spanIdToSpanMap) {
+      ListMultimap<T, Span> listMultimap = ArrayListMultimap.create();
+      for (Entry<T, List<String>> entry : sourceToSpanIdsMap.entrySet()) {
```

Comment:
Everything else was functional, why the for loop here? Multimaps work well with collectors. Similarly, try to avoid the map<T, List> and build that as a multimap.

Once you get comfortable with them, multimaps are typically easier to work with and more readable than maps holding collections. Here, you'd end up with something like
```java
return sourceToSpanIdsMultimap.entries().stream()
  .filter(entry -> spanIdToSpanMap.containsKey(entry.getValue())
  .collect(ImmutableListMultimap.toImmutableListMultimap(Entry::getKey, entry -> spanIdToSpanMap.get(entry.getValue());
```